Peg-out acceptance (#350)
rsksmart/RSKIPs: IPs/RSKIP350.md

created: 2022-09-06
---
# Peg-out Acceptance

|RSKIP          | 350          |
| :------------ |:-------------|
|**Title**      |Peg-out Acceptance|
|**Created**    |06-SEP-2022 |
|**Author**     |SDL |
|**Purpose**    |Usa |
|**Layer**      |Core |
|**Complexity** |2 |
|**Status**     |Draft |
|**discussions-to**     ||

# **Abstract**

This RSKIP proposes enables pegnatories to accept specific peg-outs from a peg-out batch, and reject specific peg-outs, or accept/reject all outstanding peg-outs.

# **Motivation**

The bridge contract is not designed to allow the blocking of peg-outs. Pegnatories may be requested to do it. To comply, pegnatories may need to turn off the PowHSM devices. However, the peg-outs are executed as soon as the devices are turned on and connected to the network again. The bridge assumes peg-out transactions are always executed, a peg-out that is already commanded cannot be cancelled. Peg-outs are grouped in batches so that a Bitcoin transaction can contain several peg-outs and therefore selective blocking cannot be performed.

In this proposal, we give pegnatories the capability to accept peg-outs before they are included in the batch and signed.

> # **Specification**

Reviewer: This section reads more like mechanics and implementation. I am not sure I follow the high level concept. Why do we need 3 actions (accept/reject/cancel)? The main idea is not very clear to me.

Author: The specification does not need to describe the high level concept. Maybe that should be explained in the Motivation section.

 
Initially, the fields `acceptanceCount`, `rejectionCount` and `voted` are set to zero. The field `previous` is set to the previous `requestId`. The field `requestEpoch` is filled with the current epoch number (block heigh divided by `epochLength`).

To help us define the external methods, we first define 3 internal methods `accept(requestId)` and `reject(requestId)`, `cancel(requestId)`.

## accept
When the `requestId` exists in the map, and the `msg.sender` is a valid pegnatory and the pegnatory index is not set in `voted`, and the reject or accept thresholds have not been reached:

1. The number of acceptances for this peg-out request (`acceptanceCount`) is incremented by one
2. The pegnatory index is set in its `voted` bit-vector.

## reject
When the `requestId` exists in the map, and the `msg.sender` is a valid pegnatory and the pegnatory index is not set in `voted`, and the reject or accept thresholds have not been reached:

1. The number of rejections for this peg-out (`rejectionCount`) request is incremented by one
2. The pegnatory index is set in its `voted` bit-vector.

## cancel
If the peg-out `requestId` request exists:

> * the storace cells used are zeroed so they are removed from storage 

Author: ```suggestion
* the storage cells used are zeroed so they are removed from storage 
```

This method is called by `updateCollections`, `releaseBtc`, `acceptPendingPegouts` and `rejectPendingPegouts`.

## acceptPendingPegouts

First, the `epochEndCheck` method is called.

This method enables the acceptance of all pending peg-out requests up to a certain pending request, and upto a maximum number of pending requests scanned given as argument. The internal `accept` method is called for each of the scanned entries.
The method does not return any value.
The gas cost of the `acceptPendingPegouts` method is set to 5000*maxEntries.

## rejectPendingPegouts

First, the `epochEndCheck` method is called.

This method enables the rejection of all pending peg-out requests up to a certain pending request, and upto a maximum number of pending requests scanned given as argument. The internal `reject` method is called for each of the scanned entries.

The method does not return any value.
The gas cost of the `rejectPendingPegouts` method is set to 5000*maxEntries.

## getPendingPegoutState

This method returns the peg-out request entry serialized. If the `requestId` does not exists, the returned data is empty.

## Internal Storage

> Entries in the `pendingPegoutRequests` are stored as independent cells in the Bridge contract storage. To derive the storage address, the message `("pendingAcceptance-" + Hex(requestId))` is hashed with Keccak256. The function Hex() returns the 64-byte lowercase hexadecimal representation of the hash without the "0x" prefix.

Reviewer 2: ```suggestion
Entries in the `pendingPegoutRequests` are stored as independent cells in the Bridge contract storage. To derive the storage address, the message `("pendingAcceptance-" + Hex(requestId))` is hashed with Keccak256. The function `Hex()` returns the 64-byte lowercase hexadecimal representation of the hash without the "0x" prefix.
```

* `requestEpoch`

The `acceptanceCount` and `rejectionCount` fields are stored as an `int16`.
The `voted ` field is a bit-vector of fixed size 16 bytes (supports a maximum of 128 pegnatories). The most significant bit of the first byte corresponds to the first pegnatory.
The `previous` is an `uint256`. All fields are fixed-length.

## Migration
During powpeg migration, when the new federation is activated, if there are peg-out request outstanding in the map then:

* All entries have their fields `acceptanceCount`, `rejectionCount` and `voted` zeroed. The fields are zeroed even if they had already reached their acceptance or rejection thresholds.
* The event `pendingPegouts(lastRequestId)` is emmitted.

The new powpeg must approve or reject the prexisting set of peg-out requests. 

## Peg-out Batching

The batch creation time is redefined in terms of epochs, specifically 3 epochs. Every 3 epochs the bridge will attempt to create a new batch. 

## Peg-out Fee Estimation

Peg-out fees will be estimated according to the number of transactions queued in the batch, as it is today, and not conting the transactions that are still in the `pendingPegoutRequests` map.

# Rationale

> This proposal is deisgned to enable batch accepts and rejections, because forcing individual votes can be very costly to pegnatories in terms of gas spent. To enalbe batch votes, it is important that the batch voted cannot change between the time the pegnatory is notified of the batch and the time the vote transaction is included in the blockchain. That's why votes apply to all peg-out requests until a specified point in the batch. An alternative would be to divide the time in two phases, one that builds a batch, and another one for voting. I nthe voting phase, new peg-outs are queued for the following batch. However, this design still has the problem that blockchain reorganizations can change the content of batches, and pegnatories votes should never mutate with block reorganizations. Therefore the votes must include either the id of a chain of peg-outs, or a block hash (which indirectly also fixes all previous peg-out requests). 

Reviewer 2: ```suggestion
This proposal is deisgned to enable batch accepts and rejections, because forcing individual votes can be very costly to pegnatories in terms of gas spent. To enalbe batch votes, it is important that the batch voted cannot change between the time the pegnatory is notified of the batch and the time the vote transaction is included in the blockchain. That's why votes apply to all peg-out requests until a specified point in the batch. An alternative would be to divide the time in two phases, one that builds a batch, and another one for voting. In the voting phase, new peg-outs are queued for the following batch. However, this design still has the problem that blockchain reorganizations can change the content of batches, and pegnatories votes should never mutate with block reorganizations. Therefore the votes must include either the id of a chain of peg-outs, or a block hash (which indirectly also fixes all previous peg-out requests). 
```

> This proposal is deisgned to enable batch accepts and rejections, because forcing individual votes can be very costly to pegnatories in terms of gas spent. To enalbe batch votes, it is important that the batch voted cannot change between the time the pegnatory is notified of the batch and the time the vote transaction is included in the blockchain. That's why votes apply to all peg-out requests until a specified point in the batch. An alternative would be to divide the time in two phases, one that builds a batch, and another one for voting. In the voting phase, new peg-outs are queued for the following batch. However, this design still has the problem that blockchain reorganizations can change the content of batches, and pegnatories votes should never mutate with block reorganizations. Therefore the votes must include either the id of a chain of peg-outs, or a block hash (which indirectly also fixes all previous peg-out requests). 

Reviewer: Typos: deisgned, enalbe,

> This proposal ensures that pegnatories will always have at least one full epoch to accept ore reject a certain peg-out request. Peg-out requests issued close to the deadline of the epoch will be moved to the next epoch. The maximum time a peg-out request can be outstanding is two epochs, minus one block. We set `epochLength` to 120, which corresponds approximately to one hour. This gives enough time for pegnatories to perform any automated check, yet it doesn't block user funds for too long. 

Reviewer: typo: "ore"

> Redefiniting the batching time in term of epochs allow to synchronize both and that the batch waiting time is not extended by one epoch. However, if a peg-out request is issued close to the end of the epoch prior a batch creation, the peg.out may not receive enough votes and it will be delayed until the net batch creation event.

Reviewer 2: ```suggestion
Redefining the batching time in term of epochs has two benefits: allows to synchronize both values, and prevents extending the batch waiting time by one epoch. However, if a peg-out request is issued close to the end of the epoch prior a batch creation, the peg.out may not receive enough votes and it will be delayed until the next batch creation event.
```
